fix: Can't upload my profile picture

### Summary

----------------------------------------------

DollyAnna, May 3, 2026, 05:21

Can't upload my profile picture

My profile picture fails to upload. I think the error code is something like 503 and it said that it failed to fulfill an apparently valid request

App Version: 1.0.9+532

### Device Information
- **platform:** android
- **model:** SM-S936B
- **manufacturer:** samsung
- **version:** 16
- **sdkInt:** 36
- **brand:** samsung

**Current Screen:** SupportCenterScreen
**User Pubkey:** npub1lxfp2fxd8ygm2grwmh7nqn42paxgmx4d69npus99urhd0c707nzqhxmhru

### Recent Logs (errors/warnings + recent context)
```
[2026-05-02T19:20:25.629449] [ERROR] (CrashReporting) Failed to initialize crash reporting: [core/duplicate-app] A Firebase App named "[DEFAULT]" already exists
[2026-05-02T19:20:25.644052] [WARNING] [VIDEO] (VideoEventService) Cannot subscribe - Nostr service not initialized (will retry when ready)
[2026-05-02T19:20:26.076950] [WARNING] [SYSTEM] Zendesk JWT: SDK not initialized
[2026-05-02T19:20:26.077041] [ERROR] [SYSTEM] (PushNotificationService) NIP-44 encryption failed for token registration
[2026-05-02T19:20:27.054160] [WARNING] [VIDEO] (VideoEventService) Default video system is disabled - users will see real content only
[2026-05-02T19:20:27.622542] [WARNING] [VIDEO] (VideoEventService) ⚠️ EOSE received but NO EVENTS for SubscriptionType.profile - feed will be empty!
[2026-05-02T19:20:27.622545] [WARNING] [VIDEO] (VideoEventService) 🔍 Running automatic diagnostics for empty SubscriptionType.profile feed...
[2026-05-02T19:20:27.622557] [WARNING] [VIDEO] (VideoEventService) 📊 Relay Status:
[2026-05-02T19:20:27.622566] [WARNING] [VIDEO] (VideoEventService)    - Configured relays: wss://relay.divine.video, wss://relay.nos.social, wss://relay.damus.io, wss://nos.lol, wss://relay.primal.net
[2026-05-02T19:20:27.622570] [WARNING] [VIDEO] (VideoEventService)    - Connected relays: wss://relay.divine.video, wss://relay.nos.social, wss://relay.damus.io, wss://nos.lol, wss://relay.primal.net (5/5)
[2026-05-02T19:20:27.622570] [WARNING] [VIDEO] (VideoEventService) 📋 Subscription Filters:
[2026-05-02T19:20:27.622572] [WARNING] [VIDEO] (VideoEventService)    Filter 0:
[2026-05-02T19:20:27.622580] [WARNING] [VIDEO] (VideoEventService)       - kinds: [34236]
[2026-05-02T19:20:27.622581] [WARNING] [VIDEO] (VideoEventService)       - authors: 1 authors
[2026-05-02T19:20:27.622590] [WARNING] [VIDEO] (VideoEventService)       - limit: 50
[2026-05-02T19:20:27.622591] [WARNING] [VIDEO] (VideoEventService)    Filter 1:
[2026-05-02T19:20:27.622592] [WARNING] [VIDEO] (VideoEventService)       - kinds: [16]
[2026-05-02T19:20:27.622593] [WARNING] [VIDEO] (VideoEventService)       - authors: 1 authors
[2026-05-02T19:20:27.622594] [WARNING] [VIDEO] (VideoEventService)       - limit: 10
[2026-05-02T19:20:27.622595] [WARNING] [VIDEO] (VideoEventService) 🔍 Testing direct database query (bypassing subscription)...
[2026-05-02T19:20:27.622668] [WARNING] [VIDEO] (VideoEventService) ⚠️ EMPTY FEED for profile:
=== EMPTY FEED DIAGNOSTIC ===
Subscription Type: profile
EOSE Duration: 572ms
Relay Connected: true
Network Online: true
Connected Relay Count: 5

Filters:
  Filter 0:
    Kinds: [34236]
    Authors: 1 authors
    First author: f9921524cd3911b5206eddfd304eaa0f4c8d9aadd1661e40a5e0eed7e3cff4c4
    Tags: 0 tags
    Since: null
    Until: null
    Limit: 50
  Filter 1:
    Kinds: [16]
    Authors: 1 authors... [truncated]
[2026-05-02T19:20:32.050784] [WARNING] [VIDEO] (VideoEventService) ✅ Direct query returned 100 video events
[2026-05-02T19:20:32.050791] [WARNING] [VIDEO] (VideoEventService) ✅ DIAGNOSTIC: Database HAS 100 events, but subscription returned 0.
[2026-05-02T19:20:32.050792] [ERROR] [VIDEO] (VideoEventService) ❌ This means subscription filtering is too restrictive OR subscription stream is broken.
[2026-05-02T19:20:32.050792] [WARNING] [VIDEO] (VideoEventService) 📄 Sample events in database:
[2026-05-02T19:20:32.050794] [WARNING] [VIDEO] (VideoEventService)    Event 0:
[2026-05-02T19:20:32.050795] [WARNING] [VIDEO] (VideoEventService)       - id: 4695b323ff68f380f49e0ebbee73ee759d788b9a48422f77fa765c17927e4836
[2026-05-02T19:20:32.050797] [WARNING] [VIDEO] (VideoEventService)       - kind: 34236
[2026-05-02T19:20:32.050798] [WARNING] [VIDEO] (VideoEventService)       - pubkey: npub1j3h3cnmvnhj00khtepd69t3gnx990k0jkxa27qyvms3wr045r76q3hfheq
[2026-05-02T19:20:32.050808] [WARNING] [VIDEO] (VideoEventService)       - createdAt: 2026-05-02 19:19:44.000
[2026-05-02T19:20:32.050811] [WARNING] [VIDEO] (VideoEventService)    Event 1:
[2026-05-02T19:20:32.050812] [WARNING] [VIDEO] (VideoEventService)       - id: 888027395f0f033745432a9467658f77a47de47a9909ec2587034f2047ed2047
[2026-05-02T19:20:32.050813] [WARNING] [VIDEO] (VideoEventService)       - kind: 34236
[2026-05-02T19:20:32.050813] [WARNING] [VIDEO] (VideoEventService)       - pubkey: npub1w08yyjy7pmxynx42m8nd4eeaqjxtxcy9ckjveck3wp2xfmkx0jnqjhl7rj
[2026-05-02T19:20:32.050817] [WARNING] [VIDEO] (VideoEventService)       - createdAt: 2026-05-02 19:19:36.000
[2026-05-02T19:20:32.050817] [WARNING] [VIDEO] (VideoEventService)    Event 2:
[2026-05-02T19:20:32.050818] [WARNING] [VIDEO] (VideoEventService)       - id: 97a38179a0e55ac6d135e9878ee0ccb57a4a688969350f803e356422d1241d97
[2026-05-02T19:20:32.050818] [WARNING] [VIDEO] (VideoEventService)       - kind: 34236
[2026-05-02T19:20:32.050819] [WARNING] [VIDEO] (VideoEventService)       - pubkey: npub17gnaelumkfa7zvrjn6yh0cgzlyg4mr33ereauj06hpcpha8hlu0qepwtsv
[2026-05-02T19:20:32.050822] [WARNING] [VIDEO] (VideoEventService)       - createdAt: 2026-05-02 19:18:17.000
[2026-05-02T19:20:32.050847] [WARNING] [VIDEO] (VideoEventService) 📊 Relay Database Stats:
[2026-05-02T19:20:32.050862] [WARNING] [VIDEO] (VideoEventService)    {connectedRelays: 5, configuredRelays: 5, relays: [wss://relay.divine.video, wss://relay.nos.social, wss://relay.damus.io, wss://nos.lol, wss://relay.primal.net]}
[2026-05-02T19:20:34.186089] [DEBUG] [AUTH] (AppRouter) Router redirect: location=/settings, authState=authenticated
[2026-05-02T19:20:34.186283] [INFO] [SYSTEM] (Route) CTX derive: type=RouteType.settings npub=null index=null
[2026-05-02T19:20:34.186327] [INFO] [VIDEO] (VideoControllerCleanup) 🧹 Route type changed (RouteType.profile → RouteType.settings), disposing all video controllers
[2026-05-02T19:20:34.186329] [DEBUG] [VIDEO] (VideoControllerCleanup) 🧹 Invalidating all video controllers
[2026-05-02T19:20:34.197089] [INFO] (ScreenAnalytics) 📱 Screen load started: settings
[2026-05-02T19:20:34.197093] [DEBUG] [UI] (PageLoadObserver) Page push tracked: settings
[2026-05-02T19:20:34.197105] [DEBUG] [VIDEO] (VideoControllerCleanup) 🛡️ Skipping duplicate invalidation (10ms since last)
[2026-05-02T19:20:34.197110] [INFO] [SYSTEM] (VideoStopNavigatorObserver) 📱 Navigation didPush to route: settings - stopped all videos
[2026-05-02T19:20:34.198205] [DEBUG] [SYSTEM] (Nip98AuthService) 📱 Creating new NIP-98 auth token for GET https://invite.divine.video/v1/invite-status
[2026-05-02T19:20:34.198307] [INFO] [AUTH] (AuthService) Signing kind 27235 via KeycastNostrIdentity (authSource=divineOAuth, eventPubkey=f9921524cd3911b5206eddfd304eaa0f4c8d9aadd1661e40a5e0eed7e3cff4c4)
[2026-05-02T19:20:34.208663] [INFO] (ProfileScreenRouter) 🧭 ProfileScreenRouter.build
[2026-05-02T19:20:34.223750] [INFO] (ScreenAnalytics) ✅ Screen content visible: settings in 26ms
[2026-05-02T19:20:34.223820] [INFO] (PagePerf) PERF: settings — visible: 26ms
[2026-05-02T19:20:34.224375] [INFO] [AUTH] (AuthService) getKnownAccounts: raw=190 chars
[2026-05-02T19:20:34.648072] [DEBUG] [SYSTEM] (RelayNotificationsProvider) RelayNotifications: Pausing auto-refresh timer (no listeners)
[2026-05-02T19:20:34.648362] [INFO] (ProfileScreenRouter) 🧭 ProfileScreenRouter.build
[2026-05-02T19:20:34.729612] [INFO] [AUTH] (AuthService) Event signed and validated: 29b02969e7b56ad96e157ba6a6b981d55c93f0350ba50889e2b33385abdfc49f
[2026-05-02T19:20:34.729707] [INFO] [SYSTEM] (Nip98AuthService) Created NIP-98 token (expires: 2026-05-02 19:30:34.729675)
[2026-05-02T19:20:34.729708] [DEBUG] [SYSTEM] (Nip98AuthService) 📱 Event ID: note19xczj608k44djms40wn2dwvp64wf8up4pwjs3z0zkvect27lcj0swhmd94
[2026-05-02T19:20:36.074258] [DEBUG] [SYSTEM] (CuratedListService) Relay sync timeout reached, processing received events
[2026-05-02T19:20:36.074373] [INFO] [SYSTEM] (CuratedListService) 📋 Received 0 raw list events from relays
[2026-05-02T19:20:36.074378] [INFO] [SYSTEM] (CuratedListService) ✅ Relay sync complete. Found 0 list events
[2026-05-02T19:20:36.074387] [INFO] [SYSTEM] (CuratedListService) Background relay sync complete, now have 1 lists
[2026-05-02T19:20:36.895583] [DEBUG] [SYSTEM] (CuratedListService) Relay sync timeout reached, processing received events
[2026-05-02T19:20:36.895723] [INFO] [SYSTEM] (CuratedListService) 📋 Received 0 raw list events from relays
[2026-05-02T19:20:36.895727] [INFO] [SYSTEM] (CuratedListService) ✅ Relay sync complete. Found 0 list events
[2026-05-02T19:20:36.895733] [INFO] [SYSTEM] (CuratedListService) Background relay sync complete, now have 1 lists
[2026-05-02T19:20:37.355557] [DEBUG] [AUTH] (AppRouter) Router redirect: location=/support-center, authState=authenticated
[2026-05-02T19:20:37.355711] [INFO] [SYSTEM] (Route) CTX derive: type=RouteType.supportCenter npub=null index=null
[2026-05-02T19:20:37.366997] [INFO] (ScreenAnalytics) 📱 Screen load started: support-center
[2026-05-02T19:20:37.367002] [DEBUG] [UI] (PageLoadObserver) Page push tracked: support-center
[2026-05-02T19:20:37.367013] [DEBUG] [VIDEO] (VideoControllerCleanup) 🧹 Invalidating all video controllers
[2026-05-02T19:20:37.367028] [INFO] [SYSTEM] (VideoStopNavigatorObserver) 📱 Navigation didPush to route: support-center - stopped all videos
[2026-05-02T19:20:37.388592] [INFO] (ScreenAnalytics) ✅ Screen content visible: support-center in 21ms
[2026-05-02T19:20:37.388690] [INFO] (PagePerf) PERF: support-center — visible: 21ms
[2026-05-02T19:20:39.258663] [DEBUG] [SYSTEM] (VideoStopNavigatorObserver) 📱 Skipping video disposal for popup route: null
[2026-05-02T19:21:52.725195] [INFO] [SYSTEM] Collecting bug report diagnostics
```

### Environment

**Platform:** 
**Build Number:** 
**OS Version:** 

### Steps to Reproduce



### Actual Result

----------------------------------------------

DollyAnna, May 3, 2026, 05:21

Can't upload my profile picture

My profile picture fails to upload. I think the error code is something like 503 and it said that it failed to fulfill an apparently valid request

App Version: 1.0.9+532

### Device Information
- **platform:** android
- **model:** SM-S936B
- **manufacturer:** samsung
- **version:** 16
- **sdkInt:** 36
- **brand:** samsung

**Current Screen:** SupportCenterScreen
**User Pubkey:** npub1lxfp2fxd8ygm2grwmh7nqn42paxgmx4d69npus99urhd0c707nzqhxmhru

### Recent Logs (errors/warnings + recent context)
```
[2026-05-02T19:20:25.629449] [ERROR] (CrashReporting) Failed to initialize crash reporting: [core/duplicate-app] A Firebase App named "[DEFAULT]" already exists
[2026-05-02T19:20:25.644052] [WARNING] [VIDEO] (VideoEventService) Cannot subscribe - Nostr service not initialized (will retry when ready)
[2026-05-02T19:20:26.076950] [WARNING] [SYSTEM] Zendesk JWT: SDK not initialized
[2026-05-02T19:20:26.077041] [ERROR] [SYSTEM] (PushNotificationService) NIP-44 encryption failed for token registration
[2026-05-02T19:20:27.054160] [WARNING] [VIDEO] (VideoEventService) Default video system is disabled - users will see real content only
[2026-05-02T19:20:27.622542] [WARNING] [VIDEO] (VideoEventService) ⚠️ EOSE received but NO EVENTS for SubscriptionType.profile - feed will be empty!
[2026-05-02T19:20:27.622545] [WARNING] [VIDEO] (VideoEventService) 🔍 Running automatic diagnostics for empty SubscriptionType.profile feed...
[2026-05-02T19:20:27.622557] [WARNING] [VIDEO] (VideoEventService) 📊 Relay Status:
[2026-05-02T19:20:27.622566] [WARNING] [VIDEO] (VideoEventService)    - Configured relays: wss://relay.divine.video, wss://relay.nos.social, wss://relay.damus.io, wss://nos.lol, wss://relay.primal.net
[2026-05-02T19:20:27.622570] [WARNING] [VIDEO] (VideoEventService)    - Connected relays: wss://relay.divine.video, wss://relay.nos.social, wss://relay.damus.io, wss://nos.lol, wss://relay.primal.net (5/5)
[2026-05-02T19:20:27.622570] [WARNING] [VIDEO] (VideoEventService) 📋 Subscription Filters:
[2026-05-02T19:20:27.622572] [WARNING] [VIDEO] (VideoEventService)    Filter 0:
[2026-05-02T19:20:27.622580] [WARNING] [VIDEO] (VideoEventService)       - kinds: [34236]
[2026-05-02T19:20:27.622581] [WARNING] [VIDEO] (VideoEventService)       - authors: 1 authors
[2026-05-02T19:20:27.622590] [WARNING] [VIDEO] (VideoEventService)       - limit: 50
[2026-05-02T19:20:27.622591] [WARNING] [VIDEO] (VideoEventService)    Filter 1:
[2026-05-02T19:20:27.622592] [WARNING] [VIDEO] (VideoEventService)       - kinds: [16]
[2026-05-02T19:20:27.622593] [WARNING] [VIDEO] (VideoEventService)       - authors: 1 authors
[2026-05-02T19:20:27.622594] [WARNING] [VIDEO] (VideoEventService)       - limit: 10
[2026-05-02T19:20:27.622595] [WARNING] [VIDEO] (VideoEventService) 🔍 Testing direct database query (bypassing subscription)...
[2026-05-02T19:20:27.622668] [WARNING] [VIDEO] (VideoEventService) ⚠️ EMPTY FEED for profile:
=== EMPTY FEED DIAGNOSTIC ===
Subscription Type: profile
EOSE Duration: 572ms
Relay Connected: true
Network Online: true
Connected Relay Count: 5

Filters:
  Filter 0:
    Kinds: [34236]
    Authors: 1 authors
    First author: f9921524cd3911b5206eddfd304eaa0f4c8d9aadd1661e40a5e0eed7e3cff4c4
    Tags: 0 tags
    Since: null
    Until: null
    Limit: 50
  Filter 1:
    Kinds: [16]
    Authors: 1 authors... [truncated]
[2026-05-02T19:20:32.050784] [WARNING] [VIDEO] (VideoEventService) ✅ Direct query returned 100 video events
[2026-05-02T19:20:32.050791] [WARNING] [VIDEO] (VideoEventService) ✅ DIAGNOSTIC: Database HAS 100 events, but subscription returned 0.
[2026-05-02T19:20:32.050792] [ERROR] [VIDEO] (VideoEventService) ❌ This means subscription filtering is too restrictive OR subscription stream is broken.
[2026-05-02T19:20:32.050792] [WARNING] [VIDEO] (VideoEventService) 📄 Sample events in database:
[2026-05-02T19:20:32.050794] [WARNING] [VIDEO] (VideoEventService)    Event 0:
[2026-05-02T19:20:32.050795] [WARNING] [VIDEO] (VideoEventService)       - id: 4695b323ff68f380f49e0ebbee73ee759d788b9a48422f77fa765c17927e4836
[2026-05-02T19:20:32.050797] [WARNING] [VIDEO] (VideoEventService)       - kind: 34236
[2026-05-02T19:20:32.050798] [WARNING] [VIDEO] (VideoEventService)       - pubkey: npub1j3h3cnmvnhj00khtepd69t3gnx990k0jkxa27qyvms3wr045r76q3hfheq
[2026-05-02T19:20:32.050808] [WARNING] [VIDEO] (VideoEventService)       - createdAt: 2026-05-02 19:19:44.000
[2026-05-02T19:20:32.050811] [WARNING] [VIDEO] (VideoEventService)    Event 1:
[2026-05-02T19:20:32.050812] [WARNING] [VIDEO] (VideoEventService)       - id: 888027395f0f033745432a9467658f77a47de47a9909ec2587034f2047ed2047
[2026-05-02T19:20:32.050813] [WARNING] [VIDEO] (VideoEventService)       - kind: 34236
[2026-05-02T19:20:32.050813] [WARNING] [VIDEO] (VideoEventService)       - pubkey: npub1w08yyjy7pmxynx42m8nd4eeaqjxtxcy9ckjveck3wp2xfmkx0jnqjhl7rj
[2026-05-02T19:20:32.050817] [WARNING] [VIDEO] (VideoEventService)       - createdAt: 2026-05-02 19:19:36.000
[2026-05-02T19:20:32.050817] [WARNING] [VIDEO] (VideoEventService)    Event 2:
[2026-05-02T19:20:32.050818] [WARNING] [VIDEO] (VideoEventService)       - id: 97a38179a0e55ac6d135e9878ee0ccb57a4a688969350f803e356422d1241d97
[2026-05-02T19:20:32.050818] [WARNING] [VIDEO] (VideoEventService)       - kind: 34236
[2026-05-02T19:20:32.050819] [WARNING] [VIDEO] (VideoEventService)       - pubkey: npub17gnaelumkfa7zvrjn6yh0cgzlyg4mr33ereauj06hpcpha8hlu0qepwtsv
[2026-05-02T19:20:32.050822] [WARNING] [VIDEO] (VideoEventService)       - createdAt: 2026-05-02 19:18:17.000
[2026-05-02T19:20:32.050847] [WARNING] [VIDEO] (VideoEventService) 📊 Relay Database Stats:
[2026-05-02T19:20:32.050862] [WARNING] [VIDEO] (VideoEventService)    {connectedRelays: 5, configuredRelays: 5, relays: [wss://relay.divine.video, wss://relay.nos.social, wss://relay.damus.io, wss://nos.lol, wss://relay.primal.net]}
[2026-05-02T19:20:34.186089] [DEBUG] [AUTH] (AppRouter) Router redirect: location=/settings, authState=authenticated
[2026-05-02T19:20:34.186283] [INFO] [SYSTEM] (Route) CTX derive: type=RouteType.settings npub=null index=null
[2026-05-02T19:20:34.186327] [INFO] [VIDEO] (VideoControllerCleanup) 🧹 Route type changed (RouteType.profile → RouteType.settings), disposing all video controllers
[2026-05-02T19:20:34.186329] [DEBUG] [VIDEO] (VideoControllerCleanup) 🧹 Invalidating all video controllers
[2026-05-02T19:20:34.197089] [INFO] (ScreenAnalytics) 📱 Screen load started: settings
[2026-05-02T19:20:34.197093] [DEBUG] [UI] (PageLoadObserver) Page push tracked: settings
[2026-05-02T19:20:34.197105] [DEBUG] [VIDEO] (VideoControllerCleanup) 🛡️ Skipping duplicate invalidation (10ms since last)
[2026-05-02T19:20:34.197110] [INFO] [SYSTEM] (VideoStopNavigatorObserver) 📱 Navigation didPush to route: settings - stopped all videos
[2026-05-02T19:20:34.198205] [DEBUG] [SYSTEM] (Nip98AuthService) 📱 Creating new NIP-98 auth token for GET https://invite.divine.video/v1/invite-status
[2026-05-02T19:20:34.198307] [INFO] [AUTH] (AuthService) Signing kind 27235 via KeycastNostrIdentity (authSource=divineOAuth, eventPubkey=f9921524cd3911b5206eddfd304eaa0f4c8d9aadd1661e40a5e0eed7e3cff4c4)
[2026-05-02T19:20:34.208663] [INFO] (ProfileScreenRouter) 🧭 ProfileScreenRouter.build
[2026-05-02T19:20:34.223750] [INFO] (ScreenAnalytics) ✅ Screen content visible: settings in 26ms
[2026-05-02T19:20:34.223820] [INFO] (PagePerf) PERF: settings — visible: 26ms
[2026-05-02T19:20:34.224375] [INFO] [AUTH] (AuthService) getKnownAccounts: raw=190 chars
[2026-05-02T19:20:34.648072] [DEBUG] [SYSTEM] (RelayNotificationsProvider) RelayNotifications: Pausing auto-refresh timer (no listeners)
[2026-05-02T19:20:34.648362] [INFO] (ProfileScreenRouter) 🧭 ProfileScreenRouter.build
[2026-05-02T19:20:34.729612] [INFO] [AUTH] (AuthService) Event signed and validated: 29b02969e7b56ad96e157ba6a6b981d55c93f0350ba50889e2b33385abdfc49f
[2026-05-02T19:20:34.729707] [INFO] [SYSTEM] (Nip98AuthService) Created NIP-98 token (expires: 2026-05-02 19:30:34.729675)
[2026-05-02T19:20:34.729708] [DEBUG] [SYSTEM] (Nip98AuthService) 📱 Event ID: note19xczj608k44djms40wn2dwvp64wf8up4pwjs3z0zkvect27lcj0swhmd94
[2026-05-02T19:20:36.074258] [DEBUG] [SYSTEM] (CuratedListService) Relay sync timeout reached, processing received events
[2026-05-02T19:20:36.074373] [INFO] [SYSTEM] (CuratedListService) 📋 Received 0 raw list events from relays
[2026-05-02T19:20:36.074378] [INFO] [SYSTEM] (CuratedListService) ✅ Relay sync complete. Found 0 list events
[2026-05-02T19:20:36.074387] [INFO] [SYSTEM] (CuratedListService) Background relay sync complete, now have 1 lists
[2026-05-02T19:20:36.895583] [DEBUG] [SYSTEM] (CuratedListService) Relay sync timeout reached, processing received events
[2026-05-02T19:20:36.895723] [INFO] [SYSTEM] (CuratedListService) 📋 Received 0 raw list events from relays
[2026-05-02T19:20:36.895727] [INFO] [SYSTEM] (CuratedListService) ✅ Relay sync complete. Found 0 list events
[2026-05-02T19:20:36.895733] [INFO] [SYSTEM] (CuratedListService) Background relay sync complete, now have 1 lists
[2026-05-02T19:20:37.355557] [DEBUG] [AUTH] (AppRouter) Router redirect: location=/support-center, authState=authenticated
[2026-05-02T19:20:37.355711] [INFO] [SYSTEM] (Route) CTX derive: type=RouteType.supportCenter npub=null index=null
[2026-05-02T19:20:37.366997] [INFO] (ScreenAnalytics) 📱 Screen load started: support-center
[2026-05-02T19:20:37.367002] [DEBUG] [UI] (PageLoadObserver) Page push tracked: support-center
[2026-05-02T19:20:37.367013] [DEBUG] [VIDEO] (VideoControllerCleanup) 🧹 Invalidating all video controllers
[2026-05-02T19:20:37.367028] [INFO] [SYSTEM] (VideoStopNavigatorObserver) 📱 Navigation didPush to route: support-center - stopped all videos
[2026-05-02T19:20:37.388592] [INFO] (ScreenAnalytics) ✅ Screen content visible: support-center in 21ms
[2026-05-02T19:20:37.388690] [INFO] (PagePerf) PERF: support-center — visible: 21ms
[2026-05-02T19:20:39.258663] [DEBUG] [SYSTEM] (VideoStopNavigatorObserver) 📱 Skipping video disposal for popup route: null
[2026-05-02T19:21:52.725195] [INFO] [SYSTEM] Collecting bug report diagnostics
```

### Expected Result



(Generated from Zendesk Ticket: rabblelabs.zendesk.com/agent/tickets/1428)

## 🔍 High-Confidence Triage Analysis

Following a log audit on #3862, we’ve identified a critical systemic blocker that links this issue to several other high-priority reports.

### 🧪 Technical Findings:
1. **Systemic Encryption Failure (NIP-44):** The logs show an explicit `[ERROR] [SYSTEM] (PushNotificationService) NIP-44 encryption failed`. This standard is required for secure identity/metadata handshakes.
2. **Identity Handshake Race Condition:** At `19:20:25-27`, the app attempts to perform secure actions (Token Registration/Upload) while the Nostr client is still in a `will retry when ready` state. This likely causes the server-side **503 error** by sending unauthenticated or malformed headers.
3. **Account Integrity:** The user is on `importedKeys`, which seems to be the common factor in several of these recent signature failures.

### 🔗 Linked Systemic Blockers (NIP-44 Cluster):
This fix is likely a "singleton" fix that will resolve:
- #3858 (Account deletion failure)
- #3743 (Push notification registration)
- #3161 (Username save failure)
- #3271 (Video deletion failure)

### 👤 Assignment & Next Steps:
**@realmeylisdev** — I am assigning this to you as the primary investigator for the NIP-44 encryption path and the identity handshake timing in the mobile app.

**@mbradley / @hm21 / @untreu2** — Please coordinate with Meylis to see if our media server/services should be hardened to provide more descriptive errors than a 503 when these handshake race conditions occur.

--- a/mobile/lib/services/upload_manager.dart
+++ b/mobile/lib/services/upload_manager.dart
@@ -2504,10 +2504,17 @@ Upload Timeout Failure:
 
       _updateUploadProgress(upload.id, 0.85);
 
-      // Upload thumbnail to Blossom server
+      // Upload thumbnail to Blossom server.
+      //
+      // `maxAttempts: 1` opts out of `BlossomUploadService.uploadImage`'s
+      // service-level retry (added for #3862). UploadManager already wraps
+      // the entire video+thumbnail pipeline in `AsyncUtils.retryWithBackoff`
+      // (see `_uploadHandler` retry config), so retrying again at the
+      // service layer would compound delays without benefit.
       final uploadResult = await _blossomService.uploadImage(
         imageFile: thumbnailFile,
         nostrPubkey: nostrPubkey,
+        maxAttempts: 1,
         onProgress: (progress) {
           // Map thumbnail progress to 85%-100% of total upload
           _updateUploadProgress(upload.id, 0.85 + (progress * 0.15));

--- a/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
+++ b/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
@@ -170,6 +170,32 @@ class BlossomUploadService {
   static const int _maxChunkRetries = 2;
   static const Duration _chunkRetryDelay = Duration(seconds: 1);
 
+  /// Default total attempts for a whole-image upload to a single server.
+  /// 1 initial + 2 retries; covers transient 5xx and connection blips on
+  /// `media.divine.video` without compounding when a caller (e.g. the
+  /// app-level UploadManager) already wraps the call in its own retry
+  /// loop — those callers pass `maxAttempts: 1`.
+  static const int _defaultUploadImageMaxAttempts = 3;
+
+  /// Base delay between retried image upload attempts. Doubled on each retry
+  /// up to [_uploadImageRetryMaxDelay].
+  static const Duration _uploadImageRetryBaseDelay = Duration(seconds: 1);
+
+  /// Hard ceiling on the per-attempt backoff for image uploads.
+  static const Duration _uploadImageRetryMaxDelay = Duration(seconds: 30);
+
+  /// HTTP status codes that justify retrying an image upload. Mirrors
+  /// `UploadManager.isRetriableError` so service-level and orchestrator-level
+  /// classification stay in sync.
+  static const Set<int> _retriableUploadStatusCodes = {
+    408, // Request timeout
+    429, // Rate limited
+    500, // Generic server error
+    502, // Bad gateway
+    503, // Service unavailable (the symptom in #3862)
+    504, // Gateway timeout
+  };
+
   /// How long a cached capability discovery result stays valid.
   static const Duration _capabilityCacheTtl = Duration(minutes: 5);
 
@@ -387,6 +413,106 @@ class BlossomUploadService {
       DioExceptionType.connectionError => true, // coverage:ignore-line
       _ => false,
     };
+  }
+
+  /// Whether an image-upload attempt threw a transient, retriable error.
+  ///
+  /// Transient = the same request stands a real chance of succeeding on a
+  /// subsequent attempt (5xx server, rate limit, connection / timeout).
+  bool _isTransientUploadError(Object error) {
+    if (error is DioException) {
+      final statusCode = error.response?.statusCode;
+      if (statusCode != null &&
+          _retriableUploadStatusCodes.contains(statusCode)) {
+        return true;
+      }
+      return switch (error.type) {
+        DioExceptionType.connectionTimeout ||
+        DioExceptionType.sendTimeout ||
+        DioExceptionType.receiveTimeout ||
+        DioExceptionType.connectionError => true,
+        _ => false,
+      };
+    }
+    return false;
+  }
+
+  /// Whether a [BlossomUploadResult] failure is transient and worth retrying.
+  ///
+  /// Some failure modes are caught inside the upload internals and surface as
+  /// `success: false` with a `statusCode`, rather than re-throwing — this
+  /// predicate covers that case symmetrically with [_isTransientUploadError].
+  bool _isTransientUploadResult(BlossomUploadResult result) {
+    if (result.success) return false;
+    final statusCode = result.statusCode;
+    return statusCode != null &&
+        _retriableUploadStatusCodes.contains(statusCode);
+  }
+
+  /// Run [attempt] up to [maxAttempts] times with exponential backoff,
+  /// stopping early on success or on a non-transient failure.
+  ///
+  /// `maxAttempts == 1` disables retry — useful for callers (UploadManager)
+  /// that already wrap the call in their own retry loop and want to avoid
+  /// compounded delays.
+  Future<BlossomUploadResult> _uploadWithRetry({
+    required Future<BlossomUploadResult> Function() attempt,
+    required int maxAttempts,
+    required String debugContext,
+  }) async {
+    var attemptNumber = 0;
+    while (true) {
+      attemptNumber++;
+      BlossomUploadResult? result;
+      Object? thrown;
+      try {
+        result = await attempt();
+      } on Object catch (e) {
+        thrown = e;
+      }
+
+      // Success path — return immediately.
+      if (result != null && result.success) {
+        return result;
+      }
+
+      // Decide whether the failure is retriable.
+      final retriable = thrown != null
+          ? _isTransientUploadError(thrown)
+          : _isTransientUploadResult(result!);
+
+      // Out of budget, or the failure mode is permanent — surface it.
+      if (attemptNumber >= maxAttempts || !retriable) {
+        if (thrown != null) {
+          // The original throwable is preserved so callers see the same
+          // exception type (DioException, BlossomResumableUploadException,
+          // etc.) they would have seen without the retry wrapper.
+          // ignore: only_throw_errors
+          throw thrown;
+        }
+        return result!;
+      }
+
+      // Compute the next backoff. Exponential: base, base*2, base*4 ...
+      // capped at the configured ceiling. attemptNumber is 1-indexed and
+      // we've just *finished* attempt N, so the multiplier is 2^(N-1).
+      final multiplier = 1 << (attemptNumber - 1);
+      final backoffMs = _uploadImageRetryBaseDelay.inMilliseconds * multiplier;
+      final cappedMs = math.min(
+        backoffMs,
+        _uploadImageRetryMaxDelay.inMilliseconds,
+      );
+      final delay = Duration(milliseconds: cappedMs);
+
+      Log.warning(
+        'Image upload attempt $attemptNumber/$maxAttempts failed '
+        '($debugContext); retrying in ${delay.inMilliseconds}ms',
+        name: 'BlossomUploadService',
+        category: LogCategory.video,
+      );
+
+      await Future<void>.delayed(delay);
+    }
   }
 
   bool _isDivineOwnedUploadHost(String serverUrl) {
@@ -1482,6 +1608,7 @@ class BlossomUploadService {
     required String nostrPubkey,
     String mimeType = 'image/jpeg',
     void Function(double)? onProgress,
+    int maxAttempts = _defaultUploadImageMaxAttempts,
   }) async {
     try {
       // Check authentication
@@ -1491,6 +1618,7 @@ class BlossomUploadService {
           errorMessage: 'Not authenticated',
         );
       }
+      assert(maxAttempts >= 1, 'maxAttempts must be at least 1');
 
       // Report initial progress
       onProgress?.call(0.1);
@@ -1539,39 +1667,51 @@ class BlossomUploadService {
 
           final capability = await _fetchDivineUploadCapability(serverUrl);
 
-          late BlossomUploadResult result;
-          if (capability.supportsResumable) {
-            result = await _uploadWithSingleLegacyFallback(
-              serverUrl: serverUrl,
-              uploadResumable: () => _uploadToServerResumable(
+          // Retry the actual upload work on transient 5xx / network blips
+          // before falling through to the next configured server. Capability
+          // discovery is not retried here — it has its own fail-soft path
+          // (`_isTransientCapabilityDiscoveryError`).
+          //
+          // #3862: previously a single 503 from the only server in the list
+          // (default config has just `media.divine.video`) surfaced to the
+          // user as a hard failure with no automatic recovery.
+          final result = await _uploadWithRetry(
+            attempt: () {
+              if (capability.supportsResumable) {
+                return _uploadWithSingleLegacyFallback(
+                  serverUrl: serverUrl,
+                  uploadResumable: () => _uploadToServerResumable(
+                    serverUrl: serverUrl,
+                    file: strippedFile,
+                    fileHash: fileHash,
+                    fileSize: fileSize,
+                    contentType: mimeType,
+                    onProgress: onProgress,
+                  ),
+                  uploadLegacyFallback: (fallbackReason) =>
+                      _uploadToServerLegacyFallback(
+                        serverUrl: serverUrl,
+                        file: strippedFile,
+                        fileHash: fileHash,
+                        fileSize: fileSize,
+                        contentType: mimeType,
+                        fallbackReason: fallbackReason,
+                        onProgress: onProgress,
+                      ),
+                );
+              }
+              return _uploadToServer(
                 serverUrl: serverUrl,
                 file: strippedFile,
                 fileHash: fileHash,
                 fileSize: fileSize,
                 contentType: mimeType,
                 onProgress: onProgress,
-              ),
-              uploadLegacyFallback: (fallbackReason) =>
-                  _uploadToServerLegacyFallback(
-                    serverUrl: serverUrl,
-                    file: strippedFile,
-                    fileHash: fileHash,
-                    fileSize: fileSize,
-                    contentType: mimeType,
-                    fallbackReason: fallbackReason,
-                    onProgress: onProgress,
-                  ),
-            );
-          } else {
-            result = await _uploadToServer(
-              serverUrl: serverUrl,
-              file: strippedFile,
-              fileHash: fileHash,
-              fileSize: fileSize,
-              contentType: mimeType,
-              onProgress: onProgress,
-            );
-          }
+              );
+            },
+            maxAttempts: maxAttempts,
+            debugContext: 'uploadImage($serverUrl)',
+          );
 
           if (result.success) {
             // Construct canonical Blossom URL from server + hash

--- a/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
+++ b/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
@@ -435,6 +435,16 @@ class BlossomUploadService {
   ///
   /// Transient = the same request stands a real chance of succeeding on a
   /// subsequent attempt (5xx server, rate limit, connection / timeout).
+  ///
+  /// Covers two thrown shapes symmetrically:
+  ///   * [DioException] with a retriable status code (5xx, 429, 408) or a
+  ///     transient connection / timeout type.
+  ///   * [BlossomResumableUploadException] with a retriable status code —
+  ///     thrown when chunk PUT exhausts its internal retry budget on a 5xx
+  ///     and the resumable session bubbles up. An outer retry rebuilds the
+  ///     session via `_initResumableUpload`. 404/410 (session expired) and
+  ///     null statusCode (auth or malformed-response) are intentionally not
+  ///     transient.
   bool _isTransientUploadError(Object error) {
     if (error is DioException) {
       final statusCode = error.response?.statusCode;
@@ -449,6 +459,11 @@ class BlossomUploadService {
         DioExceptionType.connectionError => true,
         _ => false,
       };
+    }
+    if (error is BlossomResumableUploadException) {
+      final statusCode = error.statusCode;
+      return statusCode != null &&
+          _retriableUploadStatusCodes.contains(statusCode);
     }
     return false;
   }

--- a/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
+++ b/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
@@ -1653,6 +1653,7 @@ class BlossomUploadService {
     void Function(double)? onProgress,
     int maxAttempts = _defaultUploadImageMaxAttempts,
   }) async {
+    assert(maxAttempts >= 1, 'maxAttempts must be at least 1');
     try {
       // Check authentication
       if (!authProvider.isAuthenticated) {
@@ -1661,7 +1662,6 @@ class BlossomUploadService {
           errorMessage: 'Not authenticated',
         );
       }
-      assert(maxAttempts >= 1, 'maxAttempts must be at least 1');
 
       // Report initial progress
       onProgress?.call(0.1);

--- a/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
+++ b/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
@@ -33,6 +33,7 @@ class BlossomUploadResult {
     this.blurhash,
     this.errorMessage,
     this.statusCode,
+    this.isTransientNetworkFailure = false,
   });
 
   /// Whether the upload succeeded.
@@ -70,6 +71,13 @@ class BlossomUploadResult {
 
   /// HTTP status code on failure.
   final int? statusCode;
+
+  /// Whether this failure was caused by a transient network condition
+  /// (connection / send / receive timeout, or connection error) that the
+  /// upload internals caught and converted to a result instead of
+  /// rethrowing. Surfaced as a retry signal symmetric with [statusCode]
+  /// for the cases where there is no HTTP status to classify on.
+  final bool isTransientNetworkFailure;
 
   /// Convenience getter for backwards compatibility.
   String? get cdnUrl => fallbackUrl ?? url;
@@ -439,11 +447,19 @@ class BlossomUploadService {
 
   /// Whether a [BlossomUploadResult] failure is transient and worth retrying.
   ///
-  /// Some failure modes are caught inside the upload internals and surface as
-  /// `success: false` with a `statusCode`, rather than re-throwing — this
-  /// predicate covers that case symmetrically with [_isTransientUploadError].
+  /// Some failure modes are caught inside the upload internals (notably the
+  /// legacy PUT path through [_uploadToServer]) and surface as
+  /// `success: false` rather than re-throwing. Two signals classify a
+  /// caught failure as transient, symmetric with [_isTransientUploadError]:
+  ///
+  ///   * `statusCode` is in the retriable set (5xx, 429, 408), or
+  ///   * `isTransientNetworkFailure` is `true` — set by [_uploadToServer]
+  ///     when it caught a [DioException] of type connection / send /
+  ///     receive timeout or connection error and converted it to a result
+  ///     (these have no HTTP status to classify on).
   bool _isTransientUploadResult(BlossomUploadResult result) {
     if (result.success) return false;
+    if (result.isTransientNetworkFailure) return true;
     final statusCode = result.statusCode;
     return statusCode != null &&
         _retriableUploadStatusCodes.contains(statusCode);
@@ -1299,24 +1315,28 @@ class BlossomUploadService {
           success: false,
           statusCode: statusCode,
           errorMessage: 'Connection timeout - check server URL',
+          isTransientNetworkFailure: true,
         );
       } else if (e.type == DioExceptionType.sendTimeout) {
         return BlossomUploadResult(
           success: false,
           statusCode: statusCode,
           errorMessage: 'Send timeout - upload too slow or connection dropped',
+          isTransientNetworkFailure: true,
         );
       } else if (e.type == DioExceptionType.receiveTimeout) {
         return BlossomUploadResult(
           success: false,
           statusCode: statusCode,
           errorMessage: 'Receive timeout - server not responding',
+          isTransientNetworkFailure: true,
         );
       } else if (e.type == DioExceptionType.connectionError) {
         return BlossomUploadResult(
           success: false,
           statusCode: statusCode,
           errorMessage: 'Cannot connect to Blossom server: $errorDetail',
+          isTransientNetworkFailure: true,
         );
       } else if (e.type == DioExceptionType.cancel) {
         return BlossomUploadResult(

--- a/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
+++ b/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
@@ -504,10 +504,12 @@ class BlossomUploadService {
       attemptNumber++;
       BlossomUploadResult? result;
       Object? thrown;
+      StackTrace? thrownStackTrace;
       try {
         result = await attempt();
-      } on Object catch (e) {
+      } on Object catch (e, stackTrace) {
         thrown = e;
+        thrownStackTrace = stackTrace;
       }
 
       // Success path — return immediately.
@@ -524,10 +526,10 @@ class BlossomUploadService {
       if (attemptNumber >= maxAttempts || !retriable) {
         if (thrown != null) {
           // The original throwable is preserved so callers see the same
-          // exception type (DioException, BlossomResumableUploadException,
-          // etc.) they would have seen without the retry wrapper.
-          // ignore: only_throw_errors
-          throw thrown;
+          // exception type and stack trace (DioException,
+          // BlossomResumableUploadException, etc.) they would have seen
+          // without the retry wrapper.
+          Error.throwWithStackTrace(thrown, thrownStackTrace!);
         }
         return result!;
       }

--- a/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
+++ b/mobile/packages/blossom_upload_service/lib/src/blossom_upload_service.dart
@@ -156,17 +156,24 @@ class _CachedCapability {
 /// Blossom BUD-01 media upload service with resumable upload support.
 class BlossomUploadService {
   /// Creates a [BlossomUploadService].
+  ///
+  /// [sleep] is the awaitable used to space out retries in
+  /// [_uploadWithRetry]. Defaults to `Future<void>.delayed`. Tests pass a
+  /// no-op so the retry-behavior group doesn't pay real wall time on each
+  /// backoff; production code should leave it at the default.
   BlossomUploadService({
     required this.authProvider,
     BlossomPerformanceMonitor? performanceMonitor,
     Dio? dio,
     String? defaultServerUrl,
     DateTime Function()? clock,
+    Future<void> Function(Duration)? sleep,
   }) : _performanceMonitor =
            performanceMonitor ?? const NoOpPerformanceMonitor(),
        dio = dio ?? Dio(),
        _defaultServerUrl = defaultServerUrl ?? defaultBlossomServer,
-       _clock = clock ?? DateTime.now;
+       _clock = clock ?? DateTime.now,
+       _sleep = sleep ?? Future<void>.delayed;
 
   static const String _blossomServerKey = 'blossom_server_url';
   static const String _useBlossomKey = 'use_blossom_upload';
@@ -215,6 +222,7 @@ class BlossomUploadService {
   final Dio dio;
   final String _defaultServerUrl;
   final DateTime Function() _clock;
+  final Future<void> Function(Duration) _sleep;
 
   /// In-memory cache of capability discovery results keyed by server URL.
   final Map<String, _CachedCapability> _capabilityCache = {};
@@ -527,7 +535,7 @@ class BlossomUploadService {
         category: LogCategory.video,
       );
 
-      await Future<void>.delayed(delay);
+      await _sleep(delay);
     }
   }
 

--- a/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
+++ b/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
@@ -4821,9 +4821,13 @@ void main() {
 
       setUp(() {
         mockDio = _MockDio();
+        // No-op sleep: assertions in this group are about attempt counts and
+        // final state, not backoff timing. Skipping the real 1s/2s waits cuts
+        // the group's wall time from ~16s to ~1s without changing semantics.
         service = BlossomUploadService(
           authProvider: mockAuthProvider,
           dio: mockDio,
+          sleep: (_) async {},
         );
       });
 

--- a/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
+++ b/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
@@ -5444,6 +5444,213 @@ void main() {
         },
       );
 
+      // Liz flagged that the retry helper's contract has to cover both
+      // signals: retriable HTTP statuses and the transient network
+      // failures that _uploadToServer catches and converts to result-typed
+      // failures with statusCode: null. The four tests below pin the
+      // default-host legacy PUT path for each transient DioExceptionType
+      // (connection / send / receive timeout, connection error). Without
+      // the isTransientNetworkFailure signal these would each short-circuit
+      // after one attempt — exactly the user-facing failure mode #3862
+      // describes when media.divine.video has a transient network blip.
+      test(
+        'retries on connectionTimeout from default-host legacy PUT',
+        () async {
+          stubAuth();
+          stubLegacyHead();
+
+          final imageFile = await tempImage('legacy_conn_timeout');
+          var attempt = 0;
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((_) async {
+            attempt++;
+            if (attempt == 1) {
+              throw DioException(
+                requestOptions: RequestOptions(path: '/upload'),
+                type: DioExceptionType.connectionTimeout,
+              );
+            }
+            return uploadOkResponse();
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isTrue);
+          expect(
+            attempt,
+            equals(2),
+            reason:
+                'connectionTimeout caught inside _uploadToServer is converted '
+                'to a result with statusCode: null, so the retry helper has '
+                'to look at isTransientNetworkFailure to retry.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'retries on connectionError from default-host legacy PUT',
+        () async {
+          stubAuth();
+          stubLegacyHead();
+
+          final imageFile = await tempImage('legacy_conn_error');
+          var attempt = 0;
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((_) async {
+            attempt++;
+            if (attempt == 1) {
+              throw DioException(
+                requestOptions: RequestOptions(path: '/upload'),
+                type: DioExceptionType.connectionError,
+              );
+            }
+            return uploadOkResponse();
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isTrue);
+          expect(attempt, equals(2));
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'retries on sendTimeout from default-host legacy PUT',
+        () async {
+          stubAuth();
+          stubLegacyHead();
+
+          final imageFile = await tempImage('legacy_send_timeout');
+          var attempt = 0;
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((_) async {
+            attempt++;
+            if (attempt == 1) {
+              throw DioException(
+                requestOptions: RequestOptions(path: '/upload'),
+                type: DioExceptionType.sendTimeout,
+              );
+            }
+            return uploadOkResponse();
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isTrue);
+          expect(attempt, equals(2));
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'retries on receiveTimeout from default-host legacy PUT',
+        () async {
+          stubAuth();
+          stubLegacyHead();
+
+          final imageFile = await tempImage('legacy_recv_timeout');
+          var attempt = 0;
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((_) async {
+            attempt++;
+            if (attempt == 1) {
+              throw DioException(
+                requestOptions: RequestOptions(path: '/upload'),
+                type: DioExceptionType.receiveTimeout,
+              );
+            }
+            return uploadOkResponse();
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isTrue);
+          expect(attempt, equals(2));
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'does NOT retry on cancel from default-host legacy PUT',
+        () async {
+          // Opposing-symmetry guard: cancel is a user action, not a
+          // transient network condition. _uploadToServer catches it but
+          // must NOT mark the result transient — otherwise we'd loop on a
+          // cancelled upload.
+          stubAuth();
+          stubLegacyHead();
+
+          final imageFile = await tempImage('legacy_cancel');
+          var attempt = 0;
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((_) async {
+            attempt++;
+            throw DioException(
+              requestOptions: RequestOptions(path: '/upload'),
+              type: DioExceptionType.cancel,
+            );
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isFalse);
+          expect(attempt, equals(1));
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
       test(
         'falls through to next server after exhausting retries on the first',
         () async {

--- a/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
+++ b/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
@@ -4622,9 +4622,14 @@ void main() {
             );
           });
 
+          // maxAttempts: 1 isolates this test to a single per-server attempt,
+          // which is the property under test (no duplicate legacy fallback
+          // *within* one attempt). The retry behavior across attempts is
+          // covered separately under "uploadImage - retry behavior".
           final result = await service.uploadImage(
             imageFile: imageFile,
             nostrPubkey: _testPublicKey,
+            maxAttempts: 1,
           );
 
           expect(result.success, isFalse);
@@ -4805,6 +4810,700 @@ void main() {
 
         await tempDir.delete(recursive: true);
       });
+    });
+
+    // Regression coverage for #3862: a single transient 5xx from the only
+    // configured server used to surface as a hard failure. The service now
+    // retries the per-server attempt with exponential backoff before falling
+    // through to the next server.
+    group('uploadImage - retry behavior', () {
+      late _MockDio mockDio;
+
+      setUp(() {
+        mockDio = _MockDio();
+        service = BlossomUploadService(
+          authProvider: mockAuthProvider,
+          dio: mockDio,
+        );
+      });
+
+      // Stock head() mock — capability discovery returns 200 with no
+      // resumable extension, so the legacy PUT path is taken.
+      void stubLegacyHead() {
+        when(
+          () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+        ).thenAnswer(
+          (_) async => Response(
+            requestOptions: RequestOptions(path: '/upload'),
+            statusCode: 200,
+            headers: Headers(),
+          ),
+        );
+      }
+
+      void stubAuth() {
+        when(() => mockAuthProvider.isAuthenticated).thenReturn(true);
+        when(
+          () => mockAuthProvider.createAndSignEvent(
+            kind: any(named: 'kind'),
+            content: any(named: 'content'),
+            tags: any(named: 'tags'),
+          ),
+        ).thenAnswer(
+          (_) async => _signedEvent(_testPublicKey, 24242, const [], 'upload'),
+        );
+      }
+
+      Future<File> tempImage(String tag) async {
+        final tempDir = await Directory.systemTemp.createTemp(
+          'image_retry_test_${tag}_',
+        );
+        return File('${tempDir.path}/image.jpg')
+          ..writeAsBytesSync([0xFF, 0xD8, 0xFF]);
+      }
+
+      DioException dioBadResponse(int statusCode) => DioException(
+        requestOptions: RequestOptions(path: '/upload'),
+        type: DioExceptionType.badResponse,
+        response: Response(
+          requestOptions: RequestOptions(path: '/upload'),
+          statusCode: statusCode,
+        ),
+      );
+
+      Response<dynamic> uploadOkResponse() => Response(
+        requestOptions: RequestOptions(path: '/upload'),
+        statusCode: 200,
+        data: {
+          'url': 'https://media.divine.video/hash',
+          'fallbackUrl': 'https://media.divine.video/hash',
+        },
+      );
+
+      test('retries on 503 and succeeds on second attempt', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('503');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          if (attempt == 1) throw dioBadResponse(503);
+          return uploadOkResponse();
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isTrue);
+        expect(attempt, equals(2));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test('retries on 504 and succeeds', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('504');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          if (attempt == 1) throw dioBadResponse(504);
+          return uploadOkResponse();
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isTrue);
+        expect(attempt, equals(2));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test('retries on 429 and succeeds', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('429');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          if (attempt == 1) throw dioBadResponse(429);
+          return uploadOkResponse();
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isTrue);
+        expect(attempt, equals(2));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test('retries on 408 and succeeds', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('408');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          if (attempt == 1) throw dioBadResponse(408);
+          return uploadOkResponse();
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isTrue);
+        expect(attempt, equals(2));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test(
+        'exhausts retries on persistent 503 and returns last error',
+        () async {
+          stubAuth();
+          stubLegacyHead();
+
+          final imageFile = await tempImage('exhaust');
+          var attempt = 0;
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((_) async {
+            attempt++;
+            throw dioBadResponse(503);
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isFalse);
+          expect(result.statusCode, equals(503));
+          expect(attempt, equals(3));
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test('does NOT retry on 401', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('401');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          throw dioBadResponse(401);
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isFalse);
+        expect(result.statusCode, equals(401));
+        expect(attempt, equals(1));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test('does NOT retry on 413 (file too large)', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('413');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          throw dioBadResponse(413);
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isFalse);
+        expect(result.statusCode, equals(413));
+        expect(attempt, equals(1));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test('does NOT retry on success', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('happy');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          return uploadOkResponse();
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+        );
+
+        expect(result.success, isTrue);
+        expect(attempt, equals(1));
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test('maxAttempts: 1 disables retry', () async {
+        stubAuth();
+        stubLegacyHead();
+
+        final imageFile = await tempImage('opt_out');
+        var attempt = 0;
+        when(
+          () => mockDio.put<dynamic>(
+            any(),
+            data: any(named: 'data'),
+            options: any(named: 'options'),
+            onSendProgress: any(named: 'onSendProgress'),
+          ),
+        ).thenAnswer((_) async {
+          attempt++;
+          throw dioBadResponse(503);
+        });
+
+        final result = await service.uploadImage(
+          imageFile: imageFile,
+          nostrPubkey: _testPublicKey,
+          maxAttempts: 1,
+        );
+
+        expect(result.success, isFalse);
+        expect(result.statusCode, equals(503));
+        expect(
+          attempt,
+          equals(1),
+          reason:
+              'maxAttempts: 1 must short-circuit retry so callers with their '
+              'own outer retry loop (UploadManager) do not double-retry.',
+        );
+
+        await imageFile.parent.delete(recursive: true);
+      });
+
+      test(
+        'retries on DioException thrown from non-Divine host resumable init',
+        () async {
+          // Non-Divine hosts re-throw resumable failures (no Divine-only
+          // legacy fallback). The retry helper must classify a DioException
+          // with a 5xx status as transient and retry, exercising the
+          // exception-typed `_isTransientUploadError` branch.
+          await service.setBlossomEnabled(true);
+          await service.setBlossomServer('https://third-party.example');
+
+          stubAuth();
+
+          final imageFile = await tempImage('non_divine_dio');
+
+          when(
+            () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(path: '/upload'),
+              statusCode: 200,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.extensions: [
+                  DivineUploadExtensions.resumableSessions,
+                ],
+              }),
+            ),
+          );
+
+          // Track init attempts per host so we can distinguish per-server
+          // retry from the multi-server fallback that runs after retries are
+          // exhausted.
+          final initAttemptsByHost = <String, int>{};
+          when(
+            () => mockDio.post<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+            ),
+          ).thenAnswer((invocation) async {
+            final url = invocation.positionalArguments.first as String;
+            if (!url.endsWith('/upload/init')) {
+              // Complete endpoint or anything else — return a generic OK so
+              // we only count init attempts in the assertion below.
+              return Response(
+                requestOptions: RequestOptions(path: url),
+                statusCode: 200,
+                data: {
+                  'url': 'https://media.divine.video/hash',
+                  'fallbackUrl': 'https://media.divine.video/hash',
+                },
+              );
+            }
+            final host = Uri.parse(url).host;
+            final n = (initAttemptsByHost[host] ?? 0) + 1;
+            initAttemptsByHost[host] = n;
+            if (host == 'third-party.example' && n == 1) {
+              throw DioException(
+                requestOptions: RequestOptions(path: '/upload/init'),
+                type: DioExceptionType.badResponse,
+                response: Response(
+                  requestOptions: RequestOptions(path: '/upload/init'),
+                  statusCode: 503,
+                ),
+              );
+            }
+            return Response(
+              requestOptions: RequestOptions(path: '/upload/init'),
+              statusCode: 201,
+              data: {
+                'uploadId': 'up_retry_dio',
+                'uploadUrl':
+                    'https://third-party.example/sessions/up_retry_dio',
+                'chunkSize': 1024,
+                'nextOffset': 0,
+                'expiresAt': 9999999999,
+              },
+            );
+          });
+
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(
+                path: '/sessions/up_retry_dio',
+              ),
+              statusCode: 204,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.uploadOffset: ['3'],
+              }),
+            ),
+          );
+
+          await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(
+            initAttemptsByHost['third-party.example'],
+            equals(2),
+            reason:
+                'A 503 DioException from non-Divine host resumable init must '
+                'be classified as transient by _isTransientUploadError so the '
+                'retry helper schedules another attempt on the same host.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'does NOT retry on DioException with non-retriable status from '
+        'non-Divine host',
+        () async {
+          await service.setBlossomEnabled(true);
+          await service.setBlossomServer('https://third-party.example');
+
+          stubAuth();
+
+          final imageFile = await tempImage('non_divine_dio_401');
+
+          when(
+            () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(path: '/upload'),
+              statusCode: 200,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.extensions: [
+                  DivineUploadExtensions.resumableSessions,
+                ],
+              }),
+            ),
+          );
+
+          final initAttemptsByHost = <String, int>{};
+          when(
+            () => mockDio.post<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+            ),
+          ).thenAnswer((invocation) async {
+            final url = invocation.positionalArguments.first as String;
+            final host = Uri.parse(url).host;
+            initAttemptsByHost[host] = (initAttemptsByHost[host] ?? 0) + 1;
+            throw DioException(
+              requestOptions: RequestOptions(path: '/upload/init'),
+              type: DioExceptionType.badResponse,
+              response: Response(
+                requestOptions: RequestOptions(path: '/upload/init'),
+                statusCode: 401,
+              ),
+            );
+          });
+
+          await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(
+            initAttemptsByHost['third-party.example'],
+            equals(1),
+            reason:
+                '401 is non-retriable, so a single DioException with that '
+                'status must short-circuit the retry helper for the failing '
+                'host.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'retries on DioException of connectionTimeout type from non-Divine '
+        'host',
+        () async {
+          // Connection timeout has no statusCode; classification falls to
+          // the DioExceptionType switch in _isTransientUploadError.
+          await service.setBlossomEnabled(true);
+          await service.setBlossomServer('https://third-party.example');
+
+          stubAuth();
+
+          final imageFile = await tempImage('non_divine_dio_timeout');
+
+          when(
+            () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(path: '/upload'),
+              statusCode: 200,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.extensions: [
+                  DivineUploadExtensions.resumableSessions,
+                ],
+              }),
+            ),
+          );
+
+          final initAttemptsByHost = <String, int>{};
+          when(
+            () => mockDio.post<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+            ),
+          ).thenAnswer((invocation) async {
+            final url = invocation.positionalArguments.first as String;
+            if (!url.endsWith('/upload/init')) {
+              return Response(
+                requestOptions: RequestOptions(path: url),
+                statusCode: 200,
+                data: {
+                  'url': 'https://media.divine.video/hash',
+                  'fallbackUrl': 'https://media.divine.video/hash',
+                },
+              );
+            }
+            final host = Uri.parse(url).host;
+            final n = (initAttemptsByHost[host] ?? 0) + 1;
+            initAttemptsByHost[host] = n;
+            if (host == 'third-party.example' && n == 1) {
+              throw DioException(
+                requestOptions: RequestOptions(path: '/upload/init'),
+                type: DioExceptionType.connectionTimeout,
+              );
+            }
+            return Response(
+              requestOptions: RequestOptions(path: '/upload/init'),
+              statusCode: 201,
+              data: {
+                'uploadId': 'up_retry_timeout',
+                'uploadUrl':
+                    'https://third-party.example/sessions/up_retry_timeout',
+                'chunkSize': 1024,
+                'nextOffset': 0,
+                'expiresAt': 9999999999,
+              },
+            );
+          });
+
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(
+                path: '/sessions/up_retry_timeout',
+              ),
+              statusCode: 204,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.uploadOffset: ['3'],
+              }),
+            ),
+          );
+
+          await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(
+            initAttemptsByHost['third-party.example'],
+            equals(2),
+            reason:
+                'connectionTimeout is transient and must be retried on the '
+                'same host by _isTransientUploadError.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'falls through to next server after exhausting retries on the first',
+        () async {
+          stubAuth();
+
+          await service.setBlossomEnabled(true);
+          await service.setBlossomServer('https://blossom.example.com');
+
+          final imageFile = await tempImage('multi_server');
+
+          when(
+            () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(path: '/upload'),
+              statusCode: 200,
+              headers: Headers(),
+            ),
+          );
+
+          final attemptsByHost = <String, int>{};
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer((invocation) async {
+            final url = invocation.positionalArguments.first as String;
+            final host = Uri.parse(url).host;
+            attemptsByHost[host] = (attemptsByHost[host] ?? 0) + 1;
+            if (host == 'blossom.example.com') {
+              throw dioBadResponse(503);
+            }
+            return uploadOkResponse();
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isTrue);
+          expect(
+            attemptsByHost['blossom.example.com'],
+            equals(3),
+            reason: 'Custom server should be tried 3 times before fallback.',
+          );
+          expect(
+            attemptsByHost['media.divine.video'],
+            equals(1),
+            reason:
+                'Default Divine server should succeed on its first attempt '
+                'after the custom server exhausts its retries.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
     });
 
     group('Bug report upload success path', () {

--- a/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
+++ b/mobile/packages/blossom_upload_service/test/src/blossom_upload_service_test.dart
@@ -5715,6 +5715,187 @@ void main() {
           await imageFile.parent.delete(recursive: true);
         },
       );
+
+      // Symmetry guard for _isTransientUploadError: a
+      // BlossomResumableUploadException whose statusCode is in the retriable
+      // set must be retried by the outer helper, the same way a
+      // DioException(503) already is. The init endpoint throws
+      // BlossomResumableUploadException directly for any non-200/201
+      // response that still passes validateStatus (4xx + 408 + 429 land
+      // here; 5xx is thrown by Dio as DioException(badResponse) instead).
+      test(
+        'retries on BlossomResumableUploadException with retriable status '
+        'from non-Divine host resumable init',
+        () async {
+          await service.setBlossomEnabled(true);
+          await service.setBlossomServer('https://third-party.example');
+
+          stubAuth();
+
+          final imageFile = await tempImage('non_divine_resumable_408');
+
+          when(
+            () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(path: '/upload'),
+              statusCode: 200,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.extensions: [
+                  DivineUploadExtensions.resumableSessions,
+                ],
+              }),
+            ),
+          );
+
+          final initAttemptsByHost = <String, int>{};
+          when(
+            () => mockDio.post<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+            ),
+          ).thenAnswer((invocation) async {
+            final url = invocation.positionalArguments.first as String;
+            if (!url.endsWith('/upload/init')) {
+              return Response(
+                requestOptions: RequestOptions(path: url),
+                statusCode: 200,
+                data: {
+                  'url': 'https://media.divine.video/hash',
+                  'fallbackUrl': 'https://media.divine.video/hash',
+                },
+              );
+            }
+            final host = Uri.parse(url).host;
+            final n = (initAttemptsByHost[host] ?? 0) + 1;
+            initAttemptsByHost[host] = n;
+            if (host == 'third-party.example' && n == 1) {
+              // 408 passes validateStatus (< 500) so init throws a
+              // BlossomResumableUploadException(statusCode: 408) rather
+              // than a DioException.
+              return Response(
+                requestOptions: RequestOptions(path: '/upload/init'),
+                statusCode: 408,
+                data: 'Request timeout',
+              );
+            }
+            return Response(
+              requestOptions: RequestOptions(path: '/upload/init'),
+              statusCode: 201,
+              data: {
+                'uploadId': 'up_resumable_408',
+                'uploadUrl':
+                    'https://third-party.example/sessions/up_resumable_408',
+                'chunkSize': 1024,
+                'nextOffset': 0,
+                'expiresAt': 9999999999,
+              },
+            );
+          });
+
+          when(
+            () => mockDio.put<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+              onSendProgress: any(named: 'onSendProgress'),
+            ),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(
+                path: '/sessions/up_resumable_408',
+              ),
+              statusCode: 204,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.uploadOffset: ['3'],
+              }),
+            ),
+          );
+
+          await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(
+            initAttemptsByHost['third-party.example'],
+            equals(2),
+            reason:
+                'BlossomResumableUploadException with statusCode 408 must be '
+                'classified transient by _isTransientUploadError so the outer '
+                'retry rebuilds the resumable session.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
+
+      test(
+        'does NOT retry on BlossomResumableUploadException with '
+        'non-retriable status (410) from non-Divine host resumable init',
+        () async {
+          // Opposing-symmetry guard: 410 (session expired / gone) is
+          // explicitly non-transient — re-creating the session via outer
+          // retry would only produce another 410. _isTransientUploadError
+          // must short-circuit it to a single attempt.
+          await service.setBlossomEnabled(true);
+          await service.setBlossomServer('https://third-party.example');
+
+          stubAuth();
+
+          final imageFile = await tempImage('non_divine_resumable_410');
+
+          when(
+            () => mockDio.head<dynamic>(any(), options: any(named: 'options')),
+          ).thenAnswer(
+            (_) async => Response(
+              requestOptions: RequestOptions(path: '/upload'),
+              statusCode: 200,
+              headers: Headers.fromMap({
+                DivineUploadHeaders.extensions: [
+                  DivineUploadExtensions.resumableSessions,
+                ],
+              }),
+            ),
+          );
+
+          final initAttemptsByHost = <String, int>{};
+          when(
+            () => mockDio.post<dynamic>(
+              any(),
+              data: any(named: 'data'),
+              options: any(named: 'options'),
+            ),
+          ).thenAnswer((invocation) async {
+            final url = invocation.positionalArguments.first as String;
+            final host = Uri.parse(url).host;
+            initAttemptsByHost[host] = (initAttemptsByHost[host] ?? 0) + 1;
+            return Response(
+              requestOptions: RequestOptions(path: '/upload/init'),
+              statusCode: 410,
+              data: 'Gone',
+            );
+          });
+
+          final result = await service.uploadImage(
+            imageFile: imageFile,
+            nostrPubkey: _testPublicKey,
+          );
+
+          expect(result.success, isFalse);
+          expect(
+            initAttemptsByHost['third-party.example'],
+            equals(1),
+            reason:
+                'A BlossomResumableUploadException with statusCode 410 must '
+                'short-circuit the retry helper — re-initing the session '
+                'would just return another 410.',
+          );
+
+          await imageFile.parent.delete(recursive: true);
+        },
+      );
     });
 
     group('Bug report upload success path', () {

